Enhanced Pulse Width and Angle-Based Control for Improved Arduino's Servo Compatibility

u1f992 opened the pull request:
I've found this library to be incredibly useful, especially since other popular libraries like Adafruit_NeoPixel also utilize PIO, resulting in conflicts with Servo functionality. I appreciate the effort put into its development.

The purpose of this pull request is to align the public methods of `_107_::Servo` more closely with those found in the Arduino environment, enhancing usability and compatibility.

- The SG90 is a commonly used servo motor, yet it operates within different minimum and maximum pulse widths compared to the default values currently set in the library. To accommodate this, I propose overloading the `attach` method to allow these values to be configurable.
- While the `writeMicroseconds` method is undoubtedly useful, the addition of a simpler `write` method would further streamline the user experience by mapping the 0-180 degree range directly to the minimum and maximum pulse widths.

My testing with the SG90 servo I possess has been somewhat limited, so I highly recommend a thorough review of these changes.

Thank you for considering this pull request and for your dedication to improving this library.

u1f992 committed:
Add overload of attach to customize min and max pulse width

u1f992 committed:
Implement write for angle-based control

u1f992 committed:
Remove trailing whitespaces

u1f992 committed:
Re-trigger workflow

aentinger commented on src/107-Arduino-Servo-RP2040.cpp:
```diff
@@ -66,6 +74,18 @@ void Servo::writeMicroseconds(uint16_t const pulse_width_us)
 
+void Servo::write(int angle)
+{
+  if (angle < 0)
```
Well, this mapping is kind of problematic. Who is to say that all servos will always only have 180° range?

u1f992 replied:
Your point is indeed valid, and I share some of your doubts. However, according to the [Arduino Reference](https://www.arduino.cc/reference/en/libraries/servo/write/), it is stated that the range is "from 0 to 180". It appears more accurate to consider this as mapping the 0-180 range to the full extent of the swing, regardless of the actual angle.

On a different note, it seems that the [Arduino Pico](https://github.com/earlephilhower/arduino-pico/blob/c095fce5b2f26b03fa0e6374e3371a1799abb5b1/libraries/Servo/src/Servo.cpp#L124-L132) interprets a value of less than 200 as representing an angle.

What changes would you prefer?

aentinger replied:
Personally I'd suggest to assume 360° and provide an overloaded constructor to define actual servo limits. The Arduino Pico approach is interesting because its more failsafe (i.e. more likely to obtain the desired result) for novice users but the libraries within this organisation target experienced users and therefore exactness and APIs semantically matching the expectation you have when calling them matter more.

u1f992 replied:
I see.

For now setting aside the special handling for Arduino-Pico, if we deviate from the rule that "in the `write` method, regardless of the actual angle, integers from 0 to 180 are mapped to a full swing," it seems to me that I cannot say anymore that it is "providing methods with the Arduino API signature," which is the purpose of this PR. Well, I also think this is not a very good API though...

My view is that it would be better to leave `write` as it is and provide a separate function to specify the "actual" angle from 0 to 360 degrees if we need. Perhaps introducing a new `writeDegree` method and an additional overload for ctor or `attach` to specify the actual full swing angle might be the way to go.

aentinger replied:
Lets see the code ;)

aentinger commented on src/107-Arduino-Servo-RP2040.cpp:
```diff
@@ -26,10 +26,18 @@ namespace _107_
+  attach(pin, DEFAULT_MIN_PULSE_WIDTH_us, DEFAULT_MAX_PULSE_WIDTH_us);
+}
+
+void Servo::attach(pin_size_t const pin, uint16_t min_pulse_width_us, uint16_t max_pulse_width_us)
```
```suggestion
void Servo::attach(pin_size_t const pin, uint16_t const min_pulse_width_us, uint16_t const max_pulse_width_us)
```
Use `const`-correctness as much as possible :wink:

aentinger reviewed (changes requested):
Thank you for your contribution :wink: However, we'll be needing to address a couple of conceptual issues before merging.

u1f992 committed:
Update src/107-Arduino-Servo-RP2040.cpp

Co-authored-by: Alexander Entinger <consulting@lxrobotics.com>

u1f992 committed:
Add constant and member variable for max servo angle

u1f992 committed:
Add an overload of attach method with max angle param

u1f992 committed:
Rename param from 'angle' to 'value'

u1f992 committed:
Add writeAngle method

u1f992 commented:
How about this? I added the third overload of `attach` which takes `max_angle`. In addition, I added a member variable `_max_angle`, declaring its initial value with constexpr, following the existing convention.

Since what is specified in `write` is no longer "angle," I changed the argument name to `value` and added a new method `writeAngle` to specify the actual angle.

u1f992 committed:
Add setMaxAngle method instead of overload

u1f992 commented:
Ah, sorry. I've changed my approach. Instead of using overloading, I'll create setter function `setMaxAngle`. The reason is that there is an overload in the Arduino-Pico's [Servo::attach](https://github.com/earlephilhower/arduino-pico/blob/c095fce5b2f26b03fa0e6374e3371a1799abb5b1/libraries/Servo/src/Servo.cpp#L77) that takes four arguments, which causes conflicts. Also, configuring with functions like setXxx feels more in line with the Arduino-like style.

aentinger commented on src/107-Arduino-Servo-RP2040.cpp:
```diff
@@ -66,6 +79,29 @@ void Servo::writeMicroseconds(uint16_t const pulse_width_us)
   pwm_set_chan_level(_slice_num, _channel, pulse_width_us);
 }
 
+void Servo::write(int value)
```
```suggestion
void Servo::write(int const value)
```

aentinger commented on src/107-Arduino-Servo-RP2040.cpp:
```diff
@@ -66,6 +79,29 @@ void Servo::writeMicroseconds(uint16_t const pulse_width_us)
+  if (value > 180)
+    value = 180;
+
+  uint16_t pulse_width_us = map(value, 0, 180, _min_pulse_width_us, _max_pulse_width_us);
```
```suggestion
  uint16_t const pulse_width_us = map(value, 0, 180, _min_pulse_width_us, _max_pulse_width_us);
```

aentinger commented on src/107-Arduino-Servo-RP2040.cpp:
```diff
@@ -66,6 +79,29 @@ void Servo::writeMicroseconds(uint16_t const pulse_width_us)
+  else if (angle > _max_angle)
+    angle = _max_angle;
+
+  uint16_t pulse_width_us = map(angle, 0, _max_angle, _min_pulse_width_us, _max_pulse_width_us);
```
```suggestion
  uint16_t const pulse_width_us = map(angle, 0, _max_angle, _min_pulse_width_us, _max_pulse_width_us);
```

aentinger commented on src/107-Arduino-Servo-RP2040.h:
```diff
@@ -32,23 +32,29 @@ class Servo
 
   void attach(pin_size_t const pin);
+  void attach(pin_size_t const pin, uint16_t min_pulse_width_us, uint16_t max_pulse_width_us);
+  void setMaxAngle(uint16_t const max_angle);
```
`write` and `writeAngle` take a `int` argument, `setMaxAngle` an `uint16_t`: I'd reason that it should be a `int` too. What do you think?

u1f992 replied:
I had overlooked it, I agree. Thanks for the review.

u1f992 committed:
Apply suggestions from code review

Co-authored-by: Alexander Entinger <consulting@lxrobotics.com>

aentinger commented on src/107-Arduino-Servo-RP2040.cpp:
```diff
@@ -66,6 +79,29 @@ void Servo::writeMicroseconds(uint16_t const pulse_width_us)
+void Servo::write(int const value)
+{
+  if (value < 0)
+    value = 0;
```
I assume this line breaks now, since `value` is `const`, however this can be circumvented by adding a new variable and assigning to this. I think its worth it over the emphasis that the input parameter is a constant, unchangeable value.
i.e.
```C++
int angle = value;
value = max(0, value);
value = min(_max_angle, value);
```

u1f992 committed:
Change max angle param and variable types to int

u1f992 committed:
Make params as const

u1f992 committed:
Refactor value range checks

u1f992 committed:
Early attachment check

aentinger reviewed (approved):
Terrific! I gladly accept this contribution! Good work @U-1F992 :bow:
